File: src/lib/profile-qr-link.ts
const LIGHCHAT_WEB_HOST = 'lighchat.online';
const LIGHCHAT_WEB_HOST_WWW = 'www.lighchat.online';

function normalizeUsernameToken(raw: string | null | undefined): string {
  return String(raw ?? '').trim().replace(/^@/, '').toLowerCase();
}

function looksLikeUserIdToken(token: string): boolean {
  const v = token.trim();
  if (!v) return false;
  if (v.startsWith('tg_') || v.startsWith('ya_')) return true;
  // Firebase UID in this project is usually >= 20 chars mixed-case/digits.
  return /^[A-Za-z0-9_-]{20,}$/.test(v);
}

export function buildProfileShareUrl(
  userId: string,
  username?: string | null
): string {
  const uid = userId.trim();
  if (!uid) return `https://${LIGHCHAT_WEB_HOST}/dashboard/contacts`;
  const normalizedUsername = normalizeUsernameToken(username);
  if (!normalizedUsername) {
    return `https://${LIGHCHAT_WEB_HOST}/dashboard/contacts/${encodeURIComponent(uid)}`;
  }
  return `https://${LIGHCHAT_WEB_HOST}/dashboard/contacts/${encodeURIComponent(normalizedUsername)}`;
}

export function buildProfileQrPayload(params: { userId: string; username?: string | null }): string {
  const uid = params.userId.trim();
  if (!uid) return '';
  const username = normalizeUsernameToken(params.username);
  const url = new URL(buildProfileShareUrl(uid, username));
  if (username) url.searchParams.set('u', username);
  return url.toString();
}

export type ProfileQrTarget = {
  userId: string | null;
  username: string | null;
};

export function extractProfileTargetFromQrPayload(payload: string): ProfileQrTarget {
  const empty: ProfileQrTarget = { userId: null, username: null };
  const raw = payload.trim();
  if (!raw) return empty;

  const compact = raw.replace(/\s+/g, '');
  if (compact.startsWith('lighchat_profile:')) {
    const uid = compact.slice('lighchat_profile:'.length).trim();
    return { userId: uid || null, username: null };
  }

  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch {
    return empty;
  }

  const queryUid = parsed.searchParams.get('uid') ?? parsed.searchParams.get('userId');
  const queryUsername = parsed.searchParams.get('u') ?? parsed.searchParams.get('username');
  const normalizedQueryUsername = normalizeUsernameToken(queryUsername);

  if (queryUid && queryUid.trim()) {
    return {
      userId: decodeURIComponent(queryUid.trim()),
      username: normalizedQueryUsername || null,
    };
  }

  const protocol = parsed.protocol.toLowerCase();
  if (protocol === 'lighchat:' && parsed.hostname.toLowerCase() === 'profile') {
    const path = parsed.pathname.replace(/^\/+/, '').trim();
    if (path) return { userId: decodeURIComponent(path), username: null };
  }

  const host = parsed.hostname.toLowerCase();
  const segments = parsed.pathname.split('/').filter(Boolean);
  if (
    (host === LIGHCHAT_WEB_HOST || host === LIGHCHAT_WEB_HOST_WWW) &&
    segments.length >= 3 &&
    segments[0] === 'dashboard' &&
    segments[1] === 'contacts'
  ) {
    const token = decodeURIComponent(segments[2] ?? '').trim();
    if (!token) return empty;
    if (looksLikeUserIdToken(token)) {
      return {
        userId: token,
        username: normalizedQueryUsername || null,
      };
    }
    return {
      userId: null,
      username: normalizeUsernameToken(token) || normalizedQueryUsername || null,
    };
  }

  if (segments.length >= 3 && segments[0] === 'contacts' && segments[1] === 'user') {
    const uid = decodeURIComponent(segments[2] ?? '').trim();
    if (uid) return { userId: uid, username: null };
  }

  return empty;
}

export function extractProfileUserIdFromQrPayload(payload: string): string | null {
  return extractProfileTargetFromQrPayload(payload).userId;
}

File: mobile/app/lib/features/chat/data/profile_qr_link.dart
const String kLighchatWebHost = 'lighchat.online';
const String kLighchatWebHostWww = 'www.lighchat.online';

String _normalizeUsernameToken(String? raw) {
  return (raw ?? '').trim().replaceFirst(RegExp(r'^@'), '').toLowerCase();
}

bool _looksLikeUserIdToken(String token) {
  final v = token.trim();
  if (v.isEmpty) return false;
  if (v.startsWith('tg_') || v.startsWith('ya_')) return true;
  return RegExp(r'^[A-Za-z0-9_-]{20,}$').hasMatch(v);
}

String buildProfileShareUrl(String userId, {String? username}) {
  final uid = userId.trim();
  if (uid.isEmpty) return 'https://$kLighchatWebHost/dashboard/contacts';
  final normalizedUsername = _normalizeUsernameToken(username);
  if (normalizedUsername.isEmpty) {
    return 'https://$kLighchatWebHost/dashboard/contacts/${Uri.encodeComponent(uid)}';
  }
  return 'https://$kLighchatWebHost/dashboard/contacts/${Uri.encodeComponent(normalizedUsername)}';
}

String buildProfileQrPayload({required String userId, String? username}) {
  final uid = userId.trim();
  if (uid.isEmpty) return '';
  final cleanUsername = _normalizeUsernameToken(username);
  final url = Uri.parse(buildProfileShareUrl(uid, username: cleanUsername));
  if (cleanUsername.isEmpty) return url.toString();
  return url
      .replace(queryParameters: <String, String>{'u': cleanUsername})
      .toString();
}

class ProfileQrTarget {
  const ProfileQrTarget({this.userId, this.username});

  final String? userId;
  final String? username;
}

ProfileQrTarget extractProfileTargetFromQrPayload(String payload) {
  final raw = payload.trim();
  if (raw.isEmpty) return const ProfileQrTarget();

  final compact = raw.replaceAll(RegExp(r'\s+'), '');
  if (compact.startsWith('lighchat_profile:')) {
    final uid = compact.substring('lighchat_profile:'.length).trim();
    return ProfileQrTarget(userId: uid.isEmpty ? null : uid);
  }

  final uri = Uri.tryParse(raw);
  if (uri == null) return const ProfileQrTarget();

  final fromQuery = uri.queryParameters['uid'] ?? uri.queryParameters['userId'];
  final usernameQuery = _normalizeUsernameToken(
    uri.queryParameters['u'] ?? uri.queryParameters['username'],
  );
  if (fromQuery != null && fromQuery.trim().isNotEmpty) {
    return ProfileQrTarget(
      userId: Uri.decodeComponent(fromQuery.trim()),
      username: usernameQuery.isEmpty ? null : usernameQuery,
    );
  }

  if (uri.scheme.toLowerCase() == 'lighchat') {
    if (uri.host.toLowerCase() == 'profile') {
      if (uri.pathSegments.isNotEmpty) {
        final uid = uri.pathSegments.first.trim();
        if (uid.isNotEmpty) {
          return ProfileQrTarget(userId: Uri.decodeComponent(uid));
        }
      }
    }
  }

  final host = uri.host.toLowerCase();
  final segments = uri.pathSegments;
  if ((host == kLighchatWebHost || host == kLighchatWebHostWww) &&
      segments.length >= 3 &&
      segments[0] == 'dashboard' &&
      segments[1] == 'contacts') {
    final token = Uri.decodeComponent(segments[2]).trim();
    if (token.isEmpty) return const ProfileQrTarget();
    if (_looksLikeUserIdToken(token)) {
      return ProfileQrTarget(
        userId: token,
        username: usernameQuery.isEmpty ? null : usernameQuery,
      );
    }
    final fromPath = _normalizeUsernameToken(token);
    return ProfileQrTarget(
      username: fromPath.isEmpty
          ? (usernameQuery.isEmpty ? null : usernameQuery)
          : fromPath,
    );
  }

  if (segments.length >= 3 &&
      segments[0] == 'contacts' &&
      segments[1] == 'user') {
    final uid = segments[2].trim();
    if (uid.isNotEmpty) {
      return ProfileQrTarget(userId: Uri.decodeComponent(uid));
    }
  }

  return const ProfileQrTarget();
}

String? extractProfileUserIdFromQrPayload(String payload) {
  return extractProfileTargetFromQrPayload(payload).userId;
}
